qiskit/providers/aer/openpulse/duffing_model_generators.py

```diff
@@ -0,0 +1,438 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2018, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+# pylint: disable=invalid-name
+
+"Helper functions for creating HamiltonianModel and PulseSystemModel objects"
+
+from warnings import warn
+from collections.abc import Iterable
+from .hamiltonian_model import HamiltonianModel
+from .pulse_system_model import PulseSystemModel
+
+
+def duffing_system_model(num_oscillators,
+                         dim_oscillators,
+                         oscillator_freqs,
+                         anharm_freqs,
+                         drive_strengths,
+                         coupling_dict,
+                         dt,
+                         freq_symbol='v',
+                         anharm_symbol='alpha',
+                         drive_symbol='r',
+                         coupling_symbol='j'):
+    """Returns a PulseSystemModel for a specified duffing oscillator system, and a dict specifying
+    ControlChannel indices for cross-resonance driving.
+
+    Single oscillators are specified by three parameters: frequency v, anharmonicity alpha, and
+    drive strength r, which enter into the Hamiltonian model via the terms:
+        pi*(2*v-alpha)*O + pi*alpha*O*O + 2*pi*r*X*D(t),
+    where O is the number operator, X is the drive operator, and D(t) is the signal of the
+    corresponding DriveChannel.
+
+    Couplings between oscillators are specified in the argument coupling_dict, with keys being
+    edges, and values being the coupling strenghts. Couplings enter the Hamiltonian model via an
+    exchange coupling term:
+        2*pi*j*(A0*C1+C0*A1),
+    where j is the coupling strength, and A0*C1+C0*A1 is the exchange coupling operator between the
+    two oscillators, where A is the annihiliation operator, and C is the creation operator.
+
+    For each pair of coupled oscillators, pulse ControlChannels are prepared for doing cross
+    resonance (CR) drives between oscillators. A CR drive on oscillators 0 with target oscillators
+    1 is modeled with the Hamiltonian term:
+        2*pi*r0*X0*U(t),
+    where r0 is the drive strength for oscillators 0, X0 is the drive operator for oscillators 0,
+    and U(t) is the signal for the CR drive. U(t) is set to have carrier frequency equal to that of
+    the target oscillators.
+
+    Indices for ControlChannels corresponding to CR drives are provided in the returned cr_idx_dict.
+
+    Args:
+        num_oscillators (int): Number of oscillators in the model
```

Comment:
Is this necessary? Can't the number of oscillators be inferred from the lenght of `oscillator_freqs`

Reply:
Yeah I can just change it to do that.